### PR TITLE
Fix type checking against `error[]` and `any[]`

### LINKED ISSUE #22410: Array of error is allowed to assigned to array of any
**Description:**
As `error` is not in `any` we should not allow this.

```ballerina
public function main() {
    var f = foo();
}

function foo () returns any[] {
    int[]|string[]|error[] k = <error[]>[error("reason")];
    error[] ea = [];
    any[] j = ea;
    return k;
}
```

**Steps to reproduce:**

**Affected Versions:**
1.2.0
**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
Proper type checking logic is not happening when trying to assign `error[]` to `any[]`.

Fixes #22410

## Approach
Type checking logic is added to filter out error type. Existing code let any type or array to be assigned to `any` type array.

## Samples
```Ballerina
error[] ea = [];
any[] j = ea;
```

```LogTalk
error: .::error_test_negative.bal:127:15: incompatible types: expected 'any[]', found 'error[]'
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -784,18 +784,19 @@ public boolean isArrayTypesAssignable(BType source, BType target, Set<TypePair>
         } else if (source.tag == TypeTags.ARRAY) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:784]: I think we can refactor this method a bit.

In L779 and L782 IMO we can call `isAssignable` instead of `isArrayTypesAssignable`.

AFAICT we can then change the first parameter of this method to a `BArrayType` (we'll probably have to deprecate this and introduce a new method). This'll reduce the number of specific checks and reuse the `isAssignable` logic, which has been tested quite a lot, more.

Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:784]: Sorry, I think I don't get this clearly?

The `isArrayTypesAssignable` is getting called by `isAssignable` logic anyway [link](https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L618).
Is there a point going for all the checks in `isAssignable` when we know eventually it will call `isArrayTypesAssignable` function anyway?

And changing the first parameter to BArrayType makes this specific right?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:784]: > Is there a point going for all the checks in `isAssignable` when we know eventually it will call `isArrayTypesAssignable` function anyway?
> 

L779 and L782 check assignability for the element types of the first level of arrays. And the element types may or may not be array types, so we can't say for sure it'll be calling `isArrayTypesAssignable`.

Consider 

```ballerina
string[] s = [];
(string|int)[] t = s;
```

For `(string|int)[] t = s;`, `isArrayTypesAssignable` gets invoked first, which then at L779 calls `isArrayTypesAssignable` again with source `string` and target `(string|int)` which eventually ends up calling `isAssignable` like in L807. Instead, we can directly call  `isAssignable` to check assignability of element types.



> And changing the first parameter to BArrayType makes this specific right?

IMO, the name `isArrayTypesAssignable` itself has already made it specific. :) This IS a specific check right?

Basically we call this method only when the source type is an array, to check if it is assignable to a specific type. 

So rhsType is always an array.

- if lhsType is also an array - return `isAssignable(rhsElemType, lhsElemType)`
- if lhsType is a union - return true if for at least one member `isArrayTypesAssignable` is true
- if lhsType is `json` - return `isAssignable(rhsElemType, lhsType)` 
- if lhsType is `anydata` - return `rhsType.isAnydata()` 
- if lhsType is `any` - return `rhsElemType.tag != TypeTags.ERROR`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -775,49 +776,26 @@ private boolean isArrayTypeAssignableToTupleType(BArrayType source, BTupleType t
...
         }
 
-        // Now both types are not array types and they have to be assignable
-        if (isAssignable(source, target, unresolvedTypes)) {
-            return true;
-        }
-
-        if (target.tag == TypeTags.UNION) {
-            return isAssignable(source, target, unresolvedTypes);
+        if ((target.tag == TypeTags.UNION) || (target.tag == TypeTags.JSON)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:794]: For unions, consider something like,

```ballerina
   int[] x = [1, 2];
   int|string[] y = x;
```

The current check will cause this assignment to pass right?

I think for unions we have to check `isArrayTypesAssignable` individually for each member and return true if at least one returns true.

```java
        if (target.tag == TypeTags.UNION) {
            for (BType memberType : ((BUnionType) target).getMemberTypes()) {
                if (isArrayTypesAssignable(source, memberType, unresolvedTypes) {
                    return true;
                }
            }
            return false;
        }
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:794]: Please add a couple of tests too?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -775,49 +776,26 @@ private boolean isArrayTypeAssignableToTupleType(BArrayType source, BTupleType t
...
+            return isAssignable(sourceElementType, target, unresolvedTypes);
         }
 
-        // In this case, lhs type should be of type 'any' and the rhs type cannot be a value type
-        return target.tag == TypeTags.ANY && !isValueType(source);
+        return (target.tag == TypeTags.ANY) && (sourceElementType.tag != TypeTags.ERROR);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:798]:
```ballerina
error[] x = [e1, e2];
any y = x;
```

This is valid. So we don't need this check.

File: stdlib/http/src/main/ballerina/src/http/resiliency/failover_client_endpoint.bal [hunk 4]
```diff
@@ -308,7 +308,7 @@ function performFailoverAction (string path, Request request, HttpOperation requ
     Response inResponse = new;
     HttpFuture inFuture = new;
     Request failoverRequest = request;
-    error?[] failoverActionErrData = [];
```
Comment (MaryamZi) [inline stdlib/http/src/main/ballerina/src/http/resiliency/failover_client_endpoint.bal:311]: @ThisaruGuruge can you verify these changes please? The previous behaviour (passing an `error?[]` where a `ClientError?[]` is expected in `populateFailoverErrorHttpStatusCodes`) was allowed due to a bug.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: The last check is incorrect, sorry, my bad.

See https://github.com/ballerina-platform/ballerina-lang/pull/22494#discussion_r412675176

File: stdlib/http/src/main/ballerina/src/http/resiliency/failover_client_endpoint.bal (hunk 4 above)
Comment (ThisaruGuruge) [reply stdlib/http/src/main/ballerina/src/http/resiliency/failover_client_endpoint.bal:311]: This should be fine.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 2 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:794]: Looks like unions are handled here.
https://github.com/ballerina-platform/ballerina-lang/pull/22494/files#diff-782094a79e6831ac57a34ad157af49c1L562

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 5]
```diff
@@ -775,49 +772,28 @@ private boolean isArrayTypeAssignableToTupleType(BArrayType source, BTupleType t
...
-        } else if (source.tag == TypeTags.ARRAY) {
-            // Only the right-hand side is an array type
 
-            // If the target type is a JSON, then element type of the rhs array
-            // should only be a JSON supported type.
-            if (target.tag == TypeTags.JSON) {
-                return isAssignable(((BArrayType) source).getElementType(), target, unresolvedTypes);
+            if (targetArrayType.size == source.size) {
+                return isAssignable(sourceElementType, targetElementType, unresolvedTypes);
             }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:786]: Shall we do
```suggestion
            if (targetArrayType.size != source.size) {
                return false;
            }
            
            return isAssignable(sourceElementType, targetElementType, unresolvedTypes);
```

instead?

Improves readabilty IMO.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java [hunk 6]
```diff
@@ -207,6 +207,20 @@ public void testAssignmentNegativeCases() {
...
+                                  "incompatible types: expected 'error[]', found 'any[]'", 100, 26);
+        BAssertUtil.validateError(resultNegative, i++,
+                                  "incompatible types: expected '(CError|LError)?[]', found 'error?[]'", 118, 19);
+        BAssertUtil.validateError(resultNegative, i++,
+                                  "incompatible types: expected '(CError|LError)?[]', found 'error?[]'", 119, 11);
+        BAssertUtil.validateError(resultNegative, i++,
+                                  "incompatible types: expected '(error|int[])', found 'error[]'", 127, 21);
+        BAssertUtil.validateError(resultNegative, i++,
+                                  "incompatible types: expected '(int|error[])', found 'error'", 132, 21);
+//        BAssertUtil.validateError(resultNegative, i++,
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java:222]: This can be removed?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 7]
```diff
@@ -564,12 +565,8 @@ private boolean isAssignable(BType source, BType target, Set<TypePair> unresolve
         }
 
         if (targetTag == TypeTags.JSON) {
-            if (sourceTag == TypeTags.JSON) {
```
[the remark this answers is not in the document]
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:567]: Already checked here right?
https://github.com/ballerina-platform/ballerina-lang/pull/22494/files/9ba99d6bf7cf3860b98b291a7e4a2d9f2827138f#diff-782094a79e6831ac57a34ad157af49c1L473

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java (hunk 6 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java:222]: yes

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 5 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:786]: I think this is subjective. Also, I will be adding more false statements to the method for this change. Think code is obvious anyway. I prefer to keep the code as it is if there is no objection.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java (hunk 6 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/AssignStmtTest.java]: yes, duplicate

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 5 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:786]: As a practice, we usually try to return early as and when possible, and minimize else/else-if blocks. 

With this change we can update the method itself to something like
```java
    private boolean isArrayTypesAssignable(BArrayType source, BType target, Set<TypePair> unresolvedTypes) {
        BType sourceElementType = source.getElementType();
        if (target.tag == TypeTags.ARRAY) {
            BArrayType targetArrayType = (BArrayType) target;
            BType targetElementType = targetArrayType.getElementType();
            if (targetArrayType.state == BArrayState.UNSEALED) {
                return isAssignable(sourceElementType, targetElementType, unresolvedTypes);
            }

            if (targetArrayType.size != source.size) {
                return false;
            }
            
            return isAssignable(sourceElementType, targetElementType, unresolvedTypes);
        }
        
        if (target.tag == TypeTags.JSON) {
            return isAssignable(sourceElementType, target, unresolvedTypes);
        }
        
        return false;
    }
```

And it's just one additional `return false` anyway right?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:786]: [This blog](https://medium.com/better-programming/why-you-need-to-stop-using-else-statements-5b1fd09dea9e) was shared sometime back.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal [hunk 8]
```diff
@@ -86,3 +86,41 @@ public function restActionResultAssignment() returns [int, int, string, string,
...
+    error testError = error("Test Error");
+    error[] errorArray = [testError];
+    any anyArray = errorArray;
+    error[] errorArrayBack = <error[]>anyArray;
+    assertEquality(errorMessage, errorArrayBack[0].reason());
+}
+
+function testAssignIntArrayToJson() {
+    int[*] intArray = [1, 2];
+    json jsonObjectOfIntArray = intArray;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal:101]: This variable name does not make sense IMO.

Shall we rename it to `jsonVar` or `jsonArray`?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal [hunk 9]
```diff
@@ -86,3 +86,41 @@ public function restActionResultAssignment() returns [int, int, string, string,
...
+    json jsonToJson = jsonObjectOfIntArray;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal:102]: Why do we need another assignment to a variable of the same type? What are we trying to test here?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal [hunk 10]
```diff
@@ -86,3 +86,41 @@ public function restActionResultAssignment() returns [int, int, string, string,
...
+    assertEquality("1 2", jsonToJson.toString());
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal:103]: Shall we assert the actual values instead of using `.toString()`?

How about
```
assertTrue(jsonVar is int[2]);

int[2] arr = <int[2]> jsonVar;
assertEquality(1, arr[0]);
assertEquality(2, arr[1]);
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal [hunk 11]
```diff
@@ -86,3 +86,41 @@ public function restActionResultAssignment() returns [int, int, string, string,
...
+}
+
+function testAssignIntOrStringArrayIntOrFloatOrStringUnionArray() {
+    int[]|string[] intOrStringArray = <int[]>[1, 2];
+    (int|float)[]|string[] intOrFloatArray = intOrStringArray;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal:108]: Again variable name doesn't reflect the actual type does it.

Given that these are small tests I'd rather go with simple names, `a` and `b`, `arr1` and `arr2`, etc.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal (hunk 9 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/assign/assign-stmt.bal]: extra line removed

File: tests/jballerina-bstring-unit-test/src/test/java/org/ballerinalang/test/error/ErrorTest.java [hunk 12]
```diff
@@ -270,10 +270,11 @@ public void testErrorNegative() {
                 "error reason is mandatory for direct error constructor", 112, 28);
         BAssertUtil.validateError(negativeCompileResult, i++,
                 "incompatible types: expected 'error', found '(error|int)'", 118, 11);
-        BAssertUtil.validateError(negativeCompileResult, i,
+        BAssertUtil.validateError(negativeCompileResult, i++,
```
[the remark this answers is not in the document]
Comment (KRVPerera) [reply tests/jballerina-bstring-unit-test/src/test/java/org/ballerinalang/test/error/ErrorTest.java:273]: removed

**Comment by KRVPerera** [comment]
Fixed code review suggestions
